File: 2021/Day3/Product of array except self/Product_of_array_except.cpp
class Solution
{
public:
    vector<int> productExceptSelf(vector<int> &nums)
    {
        vector<int> left(nums.size(), 1);
        vector<int> right(nums.size(), 1);
        int n = nums.size();

        for (int i = 1; i < nums.size(); i++)
        {
            left[i] = left[i - 1] * nums[i - 1];
            right[n - 1 - i] = right[n - 1 - i + 1] * nums[n - 1 - i + 1];
        }
        for (int i = 0; i < nums.size(); i++)
        {
            right[i] *= left[i];
        }
        return right;
    }
};

File: 2021/Day3/Product of array except self/Product_of_array_except.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        result = [0]*len(nums)

        prod = 1
        num0 = 0
        idx_0 = -1

        for cnt, nm in enumerate(nums):
            if(nm is 0):
                num0 += 1
                idx_0 = cnt
            elif(num0 > 1):
                break
            else:
		#Store partial products into result array
                result[cnt] = prod
                prod *= nm

        if(num0 > 1):
		#This allocate a new array instead one might want
		#to loop over result instead
            return [0]*len(nums)
        elif(num0 == 1):
		#This allocate a new array to save space one might want
		#to loop over result instead
            result = [0]*len(nums)
            result[idx_0] = prod
        else:
            prod = 1
	#Loop into reverse order and multiply partial product
	#by products computed from end
            for cnt in range(len(nums)-1, -1, -1):
                result[cnt] *= prod
                prod *= nums[cnt]

        return result
